### PR TITLE
WIP: Adam optimizer

### PR DESCRIPTION
**Pull request by ASalvail**
@MarcCote could you review and give me your thoughts? This optimizer involves a lot more updates than previous ones, but I think the logic is sound.
Will add tests in an upcoming commit.

### CODE REVIEW

File: tests/optimizers/tests_optimizers.py [hunk 1]
```diff
@@ -105,6 +99,36 @@ def test_adagrad():
...
+    # Create an Nd gaussian functions to optimize. These functions are not
+    # well-conditioned and there exists no perfect gradient step to converge in
+    # only one iteration.
+    for N in range(1, 5):
+        center = 5*np.ones((1, N)).astype(floatX)
+        param = sharedX(np.zeros((1, N)))
+        cost = T.sum(0.5*T.dot(T.dot((param-center), np.diag(1./np.arange(1, N+1))), ((param-center).T)))
+        loss = DummyLossWithGradient(cost, param)
+
+        # Even with a really high gradient step, AdaGrad can still converge.
```
Comment (MarcCote) [inline tests/optimizers/tests_optimizers.py]: Is it the case with Adam? If so, 'AdaGrad' -> Adam.

File: smartlearner/optimizers/adam.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
+from collections import OrderedDict
+
+import theano
+import theano.tensor as T
+import numpy as np
+
+from ..interfaces import Optimizer
+from ..utils import sharedX
+
+class Adam(Optimizer):
```
Comment (MarcCote) [inline smartlearner/optimizers/adam.py:9]: Why not subclass `SGD` like in `AdaGrad`? Oh wait, ok you really need the gradient. The opposite of the gradient cannot do since `mt` is signed, is that it?

File: smartlearner/optimizers/adam.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+        self.b2 = b2
+        self.eps = epsilon
+        self.lmbda = lmbda
+
+        self._updates = OrderedDict()
+        self.t = sharedX(0, name='t')
+        self.mts = []
+        self.vts = []
+
+        for i, param in enumerate(loss.gradients):  # iterates on the parameters key
```
Comment (MarcCote) [inline smartlearner/optimizers/adam.py]: Should we do that in `AdaGrad` too? Instead of creating the shared variable in the method `_get_directions`. I think the main reason it was in `_get_directions` is because `AdaGrad` was a `direction_modifier` back then instead of an `Optimizer`. We didn't have a loss object so we had to wait for the directions first before creating the shared variables.

File: smartlearner/optimizers/adam.py [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+import theano
+import theano.tensor as T
+import numpy as np
+
+from ..interfaces import Optimizer
+from ..utils import sharedX
+
+class Adam(Optimizer):
+    def __init__(self, loss, learning_rate=0.0001, b1=0.9, b2=0.999, epsilon=1e-8, lmbda=(1 - 1e-8)):
```
Comment (MarcCote) [inline smartlearner/optimizers/adam.py]: `lmbda` isn't use in the algo.

Comment (MarcCote) [inline smartlearner/optimizers/adam.py]: Also, do you know what `b1`, `b2` and the `epsilon` do? Should add short docstring for them (I know we should do it for `AdaGrad` too :P).

File: smartlearner/optimizers/adam.py [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+        self.eps = epsilon
+        self.lmbda = lmbda
+
+        self._updates = OrderedDict()
+        self.t = sharedX(0, name='t')
+        self.mts = []
+        self.vts = []
+
+        for i, param in enumerate(loss.gradients):  # iterates on the parameters key
+            self.mts.append(sharedX(np.zeros_like(param.get_value()), name='mt{}'.format(i)))
```
Comment (MarcCote) [inline smartlearner/optimizers/adam.py]: For clarity, the name of the shared variables be related to the name of the param instead of using a number. Since we already made sure that a shared variable has a valid `.name`, we could do: `name='mt_{}'.format(param.name)`.

With that you don't need the `i` variable anymore.

File: smartlearner/optimizers/adam.py [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+    def _get_directions(self):
+        directions = OrderedDict()
+        for (param, gradient), mtm1, vtm1 in zip(self.loss.gradients.items(), self.mts, self.vts):
+            tp1 = self.t + 1
+            self._updates[self.t] = tp1
+            mt = self.b1 * mtm1 + (1 - self.b1) * gradient
+            vt = self.b2 * vtm1 + (1 - self.b2) * gradient**2
+            self._updates[mtm1] = mt
+            self._updates[vtm1] = vt
+            at = self.lr * T.sqrt(1 - self.b2**tp1) / (1 - self.b1**tp1)
```
Comment (MarcCote) [inline smartlearner/optimizers/adam.py:75]: Is there a reason you didn't compute the `mt_hat` and `vt_hat` separately?

``` python
mt_hat = mt / (1 - self.b1** tp1)
vt_hat = vt / (1 - self.b2** tp1)
directions[params] = -self.lr * mt_hat / (T.sqrt(vt_hat) + self.eps)
```

In fact the way you written it is not as in the paper. You epsilon is divided by `sqrt(1-\beta_2^t)`. I don't think this has a big impact though.

File: smartlearner/optimizers/adam.py (hunk 2 above)
Comment (ASalvail) [inline smartlearner/optimizers/adam.py:9]: Actually, `AdaGrad` should subclass `Optimizer` too. It doesn't use anything from the `SGD` class.

Comment (ASalvail) [inline smartlearner/optimizers/adam.py:9]: I think it'd have been clearer to directly use `-gparam` in `AdaGrad`. In any cases, it'd make the code for Adam much more complex than it needs to be.

File: smartlearner/optimizers/adam.py (hunk 3 above)
Comment (ASalvail) [inline smartlearner/optimizers/adam.py]: If it makes the code simple to read, why not. As it is, I don't get why half of the things in AdaGrad are done. I wonder who accepted that PR !

File: smartlearner/optimizers/adam.py (hunk 4 above)
Comment (ASalvail) [inline smartlearner/optimizers/adam.py]: Good catch. the `epsilon` is for numerical stability (prevents div by 0). `b1` and `b2` are decay of the learning rate.

I'll add some details to the function.

File: smartlearner/optimizers/adam.py (hunk 5 above)
Comment (ASalvail) [inline smartlearner/optimizers/adam.py]: I like the idea. I'm a bit scared there'll be a lot of `NoneType` errors, though.

File: smartlearner/optimizers/adam.py (hunk 6 above)
Comment (ASalvail) [inline smartlearner/optimizers/adam.py:75]: This form is documented in the article at the end of the section 2.

Comment (ASalvail) [inline smartlearner/optimizers/adam.py:75]: I don't get the part about the epsilon...

File: smartlearner/optimizers/adam.py (hunk 2 above)
Comment (MarcCote) [inline smartlearner/optimizers/adam.py:9]: `AdaGrad` uses the _get_directions of `SGD that is it works with the opposite direction of the gradient.

File: smartlearner/optimizers/adam.py (hunk 3 above)
Comment (MarcCote) [inline smartlearner/optimizers/adam.py]: Apart from the initialisation of the shared variables what is it that you don't understand?

File: smartlearner/optimizers/adam.py (hunk 5 above)
Comment (MarcCote) [inline smartlearner/optimizers/adam.py]: Shouldn't be the case since we already make sure `my_shared_variable.name` is either the name or auto_name. That said, it is true that's imply the use of our `sharedX` function :S.

File: smartlearner/optimizers/adam.py (hunk 2 above)
Comment (ASalvail) [inline smartlearner/optimizers/adam.py:9]: Which would have been as well with an iterator of the gradients and the use of the opposite gradients. I'm not even sure you'd save on LOCs.

File: smartlearner/optimizers/adam.py (hunk 5 above)
Comment (ASalvail) [inline smartlearner/optimizers/adam.py]: Where? In sharedX, we use `None` quite clearly.

File: smartlearner/optimizers/adam.py (hunk 6 above)
Comment (MarcCote) [inline smartlearner/optimizers/adam.py:75]:
``` latex
at = lr * \sqrt(1 - \beta_2^t) / (1 - \beta_1^t)
direction = -at * mt / (\sqrt(vt) + \epsilon)
               = -(lr * \sqrt(1 - \beta_2^t) / (1 - \beta_1^t)) * mt / (\sqrt(vt) + \epsilon)
               = -lr * ( mt / (1 - \beta_1^t) ) * ( \sqrt(1 - \beta_2^t) / (\sqrt(vt) + \epsilon) )
               = -lr * mt_hat / ( (\sqrt(vt) + \epsilon) / \sqrt(1 - \beta_2^t) )
               = -lr * mt_hat / ( vt_hat + \epsilon / \sqrt(1 - \beta_2^t) )
```

Comment (ASalvail) [inline smartlearner/optimizers/adam.py:75]: That explains why they used `\hat\epsilon`...

File: smartlearner/optimizers/adam.py (hunk 2 above)
Comment (MarcCote) [inline smartlearner/optimizers/adam.py:9]: I agree, it probably doesn't matter. And subclassing directly `Optimizer` has the advantage of being clearer (i.e. that we use the opposite of the gradient). Like I said `AdaGrad`'s code dates from when it was implemented as an `DirectionModifier`/`UpdateRule`. It is fine with me if we adapt it to better fit the current framework.

File: smartlearner/optimizers/adam.py (hunk 6 above)
Comment (MarcCote) [inline smartlearner/optimizers/adam.py:75]: I see.

File: smartlearner/optimizers/adam.py (hunk 5 above)
Comment (MarcCote) [inline smartlearner/optimizers/adam.py]: Oh, my mistake, I was thinking of `MonitorVariable`. Ok, I would still prefer having a meaningful name, even it requires to do the same tricks as in `MonitorVariable` (i.e. use `auto_name` if `name` is `None`).

Comment (ASalvail) [inline smartlearner/optimizers/adam.py]: We could change the way sharedX works though.

**Comment by ASalvail** [comment]
That'd be ready to merge!

File: smartlearner/optimizers/adam.py (hunk 5 above)
Comment (MarcCote) [inline smartlearner/optimizers/adam.py]: Yes, we could make the `name` parameter of `sharedX`not optional anymore.
